src/Perpetuum/Services/EventServices/EPBonusEventService.cs

```diff
@@ -0,0 +1,60 @@
+﻿using Perpetuum.Accounting;
+using Perpetuum.Threading.Process;
+using System;
+using System.Threading.Tasks;
+
+namespace Perpetuum.Services.EventServices
+{
+    public class EPBonusEventService : Process
+    {
+        private IAccountManager _accountManager;
+        private TimeSpan _duration; 
+        private TimeSpan _elapsed; 
+        private bool _eventStarted; 
+        private bool _endingEvent; 
+
+        public EPBonusEventService(IAccountManager accountManager)
+        {
+            _accountManager = accountManager;
+            _eventStarted = false;
+            _endingEvent = false;
+            _duration = TimeSpan.MaxValue;
+            _elapsed = TimeSpan.Zero;
+        }
+
+        public void SetEvent(int bonus, TimeSpan duration)
+        {
+            _duration = duration;
+            _elapsed = TimeSpan.Zero;
+            _accountManager.SetEPBonusBoost(bonus);
+            _eventStarted = true;
+        }
+
+        private void DoClearBonus()
+        {
+            _accountManager.SetEPBonusBoost(0);
+        }
+
+        public override void Update(TimeSpan time)
+        {
+            if (!_eventStarted)
+                return;
+
+            if (_endingEvent)
+                return;
+
+            _elapsed += time;
+
+            if (_elapsed < _duration)
+                return;
+
+            _endingEvent = true;
+            Task.Run(() => DoClearBonus()).ContinueWith(t =>
```

Comment:
Try and avoid using this pattern; it can behave unexpectedly. For example in MVC, the ContinueWith would never execute because there is a change of SynchronisationContext between the executor and the target. Or in shared TargetSynchronisationContexts (typically WinForms apps) you can deadlock as it waits to join the thread that's waiting for it to join.

It seems like you want Update to finish executing (so as to not hang up the Server event loop) but you want DoClearBonus and the variables to be executed synchronously.

I would in this case, simply declare the entire block within the Run statement, or, create a method and call that.

eg. 
```
Task.Run(() =>
{
    DoClearBonus();
    _eventStart = false;
    _endingEvent = false;
    _elapsed = TimeSpan.Zero;
});
```
If you don't care about the synchronous execution of `DoClearBonus()` then you can make the Runner asynchronous.
```
Task.Run(async () =>
{
    await Task.Run(() => DoClearBonus());
    _eventStart = false;
    _endingEvent = false;
    _elapsed = TimeSpan.Zero;
});
```